### PR TITLE
Use Optim instead of home made Newton-Raphson

### LINKED ISSUE #11: Use Optim or similar for Newton-Rhapson
Rather than rolling our own optimization for Cox models, I think we might be better off depending on a well-established package such as Optim to perform that if at all possible.

### LINKED ISSUE #33: Sporadic CI failures in Cox regression
E.g. https://github.com/JuliaStats/Survival.jl/runs/3504403977. It happens randomly on both 1.6 and nightly although there shouldn't be anything stochastic about that test. I have no idea what is causing this and I can't reproduce locally.

### PR DESCRIPTION
**Pull request by andreasnoack**
Fixes #11 and hopefully fixes #33

### CODE REVIEW

File: src/cox.jl [hunk 1]
```diff
@@ -123,65 +123,83 @@ StatsBase.stderror(obj::CoxModel) = sqrt.(diag(vcov(obj)))
...
+_coxph(X::AbstractArray{<:Integer}, s::AbstractVector; tol, l2_cost) = _coxph(float(X), s; tol, l2_cost)
 
-function _coxph(X::AbstractArray{T}, s::AbstractVector; l2_cost=zero(T), kwargs...) where T
+function _coxph(X::AbstractArray{T}, s::AbstractVector; l2_cost, tol) where T
     R = promote_nonmissing(T)
     c = CoxAux(X, s, l2_cost)
-    fgh! = (β, grad, hes, compute_derivatives)->_cox_fgh!(β, grad, hes, c, compute_derivatives)
-    β, neg_ll, grad, hes = newton_raphson(fgh!, zeros(R, size(X, 2)); kwargs...)
-    CoxModel{R}(c, β, -neg_ll, -grad, hes, pinv(hes))
+    β₀ = randn(R, size(X, 2))
```
Comment (ararslan) [inline src/cox.jl:198]: I haven't thought about optimization in a long time. Why do we need to randomly initialize the betas? Seems we were initializing to zero before. Is this related to #33, like the search is just getting stuck based on where it started?

File: src/cox.jl [hunk 2]
```diff
@@ -193,11 +211,11 @@ Given a matrix `M` of predictors and a corresponding vector of events, compute t
 Cox proportional hazard model estimate of coefficients. Returns a `CoxModel`
 object.
 """
-function StatsBase.fit(::Type{CoxModel}, M::AbstractMatrix, y::AbstractVector; kwargs...)
+function StatsBase.fit(::Type{CoxModel}, M::AbstractMatrix, y::AbstractVector; tol=1e-4, l2_cost=0)
     index_perm = sortperm(y)
     X = M[index_perm,:]
     s = y[index_perm]
-    _coxph(X, s; kwargs...)
+    _coxph(X, s; tol, l2_cost)
```
Comment (ararslan) [inline src/cox.jl:218]:
```suggestion
    _coxph(X, s; tol=tol, l2_cost=l2_cost)
```
I guess here too for 1.0 support.

**Review by ararslan — COMMENTED** [review]
Broadly looks good to me, I think. (I haven't thought about the internals here in quite some time.) I just have a couple of comments and a question.

The CI failure on nightly seems like a difference in precision that we could likely fix by loosening the tolerance for that test.

File: src/cox.jl (hunk 1 above)
Comment (andreasnoack) [reply src/cox.jl]: This was a mistake. I changed it to `randn` during the transition to make it easier to detect errors in the objective function. I've reverted the change.

**Comment by andreasnoack** [comment]
I've addressed the comments but also bumped the minimum Julia version to 1.2 to avoid very old versions of Optim which made the results slightly off.

**Review by ararslan — APPROVED** [review]
:sparkling_heart: Thanks for doing this!

**Pull request merged**